### DISCUSSION TITLE
how to store websocket(object) connection in memcache?

### DISCUSSION THREAD

**Discussion by akshaybhuradia2020** [fastapi/fastapi#12124] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.websocket("/ws/{client_id}")
async def websocket_endpoint(
    _websocket: WebSocket,
    client_id: str
):
    try:
        serialized_obj = pickle.dumps(obj=_websocket, protocol=pickle.HIGHEST_PROTOCOL)
        memcache_client.set_cache(key=f"ws_{client_id}", value=serialized_obj)
        connections : WebSocket = pickle.loads(memcache_client.get_cache(key=f"ws_{client_id}"))
        while True:
           '''logic'''

    except WebSocketDisconnect:
        await connections.close(code=1000, reason="socket error")
```

### Description

I want to store websocket object in memcache db. but i am getting error related to serialization.
```
    serialized_obj = pickle.dumps(obj=_websocket, protocol=pickle.HIGHEST_PROTOCOL)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: Can't pickle local object 'FastAPI.setup.<locals>.openapi'
INFO:     connection open
INFO:     connection closed
```

### Operating System

Linux

### FastAPI Version

0.112.2

### Pydantic Version

2.8.2

### Python Version

Python 3.11.8

**Comment by YuriiMotov** [comment]
Could you explain your purpose of doing that?

**Reply by akshaybhuradia2020** [reply]
USECASE : to give notification to web application. after user signup, a 3rd party webhook gives user verification status. so when i get response from webhook. i give notification or send message to particular user , a verfication status. so i thought to creating websocket connection and when server give response, i ll send notifcation about status and  disconnect that particular user.
purpose is to track user:
1. when user connect first time. i ll store connection in memcache.
2. send msg to specific user only.
3. disconnect user when user get approval.

**Reply by YuriiMotov** [reply, marked as answer]
You haven't mentioned this detail, but I suppose you want to share these websocket objects between workers (otherwise you could just store them in global `dict` object).
If so, you will not be able to do this.

I don't quite understand your task. But in general, you need some kind of distributed queue (you can use Redis Pub/Sub, RabbitMQ Publish/Subscribe, or something similar) to connect websocket endpoint with other endpoints.
When in one of your endpoints you need to send notification via websocket, you need to write this message into that queue.
Websocket endpoint should monitor that queue (`while not_disconnected` loop) and if there is a notification to send, it should get this notification from the queue and send to the client via opened websocket connection.

**Reply by akshaybhuradia2020** [reply]
my task is to send message from other part of (or from other module)code based on active websocket connection object, which was created when user is connected first time.

**Reply by YuriiMotov** [reply]
> my task is to send message from other part of (or from other module)code based on active websocket connection object, which was created when user is connected first time.

If that part of code is executed in the same event loop, then you don't need memcache (use global `dict` instead).
But if that part of code is executed in another process, you can't just pass `WebSocket` object to that process and use it there. In this case use the approach I described above
